source/_posts/2020-07-28-Kubernetes-controllers-A-New-Hope.md

```diff
@@ -0,0 +1,297 @@
+
+## Kubernetes controllers today
+
+Today a Kubernetes controller consists in more or less:
+
+* A main loop that does long polling to Kubernetes to watch one or more resources
+* On every event, a function called _reconciler_ is invoked to align the cluster state to the desired state. For example: if the desired state expects a Pod to be running, but for some reason there isn't, then the _reconciler_ creates it.
+* The _reconciler_ optionally set the status of the CR, like _Ready_, _SomeError_, ...
+
+After we implement the controller, we need to configure and apply the _Custom Resource Definition_ and we need to setup a _Service Account_, that is an account to connect to our Kubernetes cluster in order to perform the operations the _reconciler_ does.
+
+During the years a lot of tooling was created, _mostly in Golang_, to implement these controllers, depending on the user needs. To mention some of them:
+
+* [Kubebuilder](https://github.com/kubernetes-sigs/kubebuilder)
+* [Operator SDK](https://sdk.operatorframework.io/)
+
+While other projects, like Knative, opt for creating and maintaining their [own tooling](https://github.com/knative/pkg/tree/master/controller) to implement controllers.
+
+These tools usually provides:
+
+* Library to improve the event handling logic: "smart" events queue that dedup events, filtering events, ...
+* A code generator component that, starting from the CRD defined in the code, scaffolds the controller code and the client code to interact with the CRD (this is because Golang doesn't have generics)
+
+Apart the differences between the tools, they all share the same problem: one CRD maps to one controller that maps to one process that maps to one pod. Every time we want to extend Kubernetes with a new API we add a new process which generates network traffic and waste CPU and memory, even when it's just listening. 
+In short, **We're flooding our Kubernetes clusters with applications that suck resources and remain in idle for most of their time!**
+
+Knative community, with some well crafted engineering, found a solution to package different controllers **at compile time** in the same process. This is fine, but that's still not a solution to the initial problem: we cannot build Kubernetes every time packaging all the controllers we need, we want a solution to dynamically extend the Kubernetes APIs.
+
+## Extending Kubernetes APIs in process
+
+The idea behind our solution is to create a runtime plugin system where every plugin is a Kubernetes controller.
+
+{%fi /images/kubernetes-controllers-a-new-hope/kubernetes-controllers-a-new-hope-1.png, Plugin system, Plugin system %}
+
+This plugin system should have certain traits:
+
+* Plugins should be well isolated between them aka no plugin failure/bad behaviour should annoy other plugins
+* The host should be able to control and restrict what the plugin can do
+* Users should be able to develop plugins with different languages
```

Comment:
We also want to dynamically load and unload plugins at runtime. Might be obvious, but it's still a trait of the system.

Reply:
added